blog/_posts/2019-03-19-debuggers.md

```diff
@@ -0,0 +1,254 @@
+with the `@enter` macro. Execution is then suspended on the first line of the function (line 4)
+and it is possible to see a breakpoint on line 8. Running the command `c` (short for "continue")
+execution is resumed until a breakpoint is encountered. At this point, the command `fr`
+(short for "frame") shows all the local variables and their values at the point
+where the execution of the code suspended due to the breakpoint.
+Finally, a "Julia REPL mode" is entered using the `` ` `` key.
+This gives a normal Julia REPL mode with the addition that the local variables are available.
+
+Rebugger enters calls via a key binding, by default Meta-i (probably
+Esc-i, Alt-i, or option-i).  Try typing `gcd(10, 20)`, and *without
+hitting enter* type Meta-i. After a short pause you should see some
+display; type `?` to see the possible actions:
+
+![Rebugger](/images/blog/2019-03-19-debuggers/rebugger_interpret.png)
+
+Rebugger also features an "edit" interface. For more information, see
+[Rebugger's documentation](https://timholy.github.io/Rebugger.jl/dev/).
+
+# An overview of the packages
+
+With several packages making their initial debut, and some old ones
+getting new capabilities, we felt it would be appropriate to provide
+an overview of the underpinnings of the new ecosystem.
+
+## JuliaInterpreter
+
+[JuliaInterpreter][JI] is the lynchpin of the entire stack; it
+contains the logic needed to evaluate and inspect running Julia code.
+An
+[interpreter](https://en.wikipedia.org/wiki/Interpreter_(computing))
+lends itself naturally to stepwise code evaluation and the
+implementation of breakpoints.
+
+JuliaInterpreter descended from an original package
+[ASTInterpreter2](https://github.com/JuliaDebug/ASTInterpreter2.jl)
+written by Keno Fischer. In its original form (prior to Jan 2019),
+ASTInterpreter2 was a fairly small but sophisticated package, capable
+of handling many advanced aspects of Julia's internal representation
+of code.  It was in need of updating to the many changes in Julia 1.0,
+although most of that work had already been done by Neethu Joy in late
+2018.  When we began our own efforts, we finished the updates and
+decided to extend it in many ways:
+
+- JuliaInterpreter became recursive-by-default, interpreting calls all
+  the way down to the `ccall`s, intrinsic functions, and builtin
+  functions that define Julia's lowest levels.  By running virtually all code
+  through the interpreter, it became more straightforward to implement
+  breakpoints and trap errors.
+
+- JuliaInterpreter received numerous performance enhancements, and now
+  can run stepwise through code at roughly 50× its original speed.
+  These optimizations reduce---but come nowhere close to
+  eliminating---the most serious disadvantage of running all code in the
+  interpreter: slow performance. It is hoped that the performance gap
+  between compiled and interpreted code, which is now as high as 1000×
```

Comment:
I think we should boost this number and emphasize this a bit more. There are cases with 1_000_000 slowdown.

Reply:
I'll go with "many orders of magnitude"